llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp

```diff
@@ -1730,13 +1736,21 @@ struct DSEState {
       // Remove its operands
       for (Use &O : DeadInst->operands())
         if (Instruction *OpI = dyn_cast<Instruction>(O)) {
-          O = nullptr;
+          O.set(PoisonValue::get(O->getType()));
           if (isInstructionTriviallyDead(OpI, &TLI))
             NowDeadInsts.push_back(OpI);
         }
 
       EI.removeInstruction(DeadInst);
-      DeadInst->eraseFromParent();
+      // Remove memory defs directly, but only queue other dead instructions for
+      // later removal. They may have been used as memory locations that have
+      // been cached by BatchAA. Removing them here may lead to newly created
+      // instructions to be allocated at the same address, yielding stale cache
+      // entries.
+      if (IsMemDef)
+        DeadInst->eraseFromParent();
```

Comment:
Not sure if there could be cases where a MemoryDef also loads a value that is then used as pointer of MemoryLocation

Reply:
+1 to this question, I'm not clear if this is possible. MemoryDefs could be volatile loads, but this is not the case here since the pass is only looking at store/writing instructions. A memcpy does also load, so a memcpy to a location that is later overwritten, could be deleted here, while copying from a location that's already in the cache.

Reply:
At this time I'm inclined to say this cannot happen. So let's have this fix in for the existing issue.

Reply:
Deleting a memcpy wouldn't be a problem, as long as its pointer operands aren't deleted. 

But we might remove something like a `malloc` call, which is a memory def, and the result of the malloc could be used as address and be in the cache. Updated the code to remove memory-defs that do not produce a value (i.e. have type void)

Reply:
Is it possible to add more tests to cover the void/non-void type cases mentioned?

Reply:
Added a couple of extra tests based on the original test.